# <font color='red'><b>Bootstrap assignment</b> </font>

<b>There will be some functions that start with the word "grader" ex: grader_sampples(), grader_30().. etc, you should not change those function definition.

Every Grader function has to return True.</b>

<font color='blue'> <b>Importing packages</b> </font>

In [ ]:
import numpy as np # importing numpy for numerical computation
from sklearn.datasets import load_boston # here we are using sklearn's boston dataset
from sklearn.metrics import mean_squared_error # importing mean_squared_error metric
import warnings
warnings.filterwarnings("ignore")

In [ ]:
boston = load_boston()
x=boston.data #independent variables
y=boston.target #target variable

In [ ]:
x.shape

(506, 13)

## <font color='red'><b>Task 1</b></font>

<font color='red'> <b>Step - 1</b></font>

*  <font color='blue'><b>Creating samples</b></font><br>
    <b> Randomly create 30 samples from the whole boston data points</b>
    *  Creating each sample: Consider any random 303(60% of 506) data points from whole data set and then replicate any 203 points from the sampled points
    
     For better understanding of this procedure lets check this examples, assume we have 10 data points [1,2,3,4,5,6,7,8,9,10], first we take 6 data points randomly , consider we have selected [4, 5, 7, 8, 9, 3] now we will replicate 4 points from [4, 5, 7, 8, 9, 3], consder they are [5, 8, 3,7] so our final sample will be [4, 5, 7, 8, 9, 3, 5, 8, 3,7]
* <font color='blue'><b> Create 30 samples </b></font>
    *  Note that as a part of the Bagging when you are taking the random samples <b>make sure each of the sample will have different set of columns</b><br>
Ex: Assume we have 10 columns[1 ,2 ,3 ,4 ,5 ,6 ,7 ,8 ,9 ,10] for the first sample we will select [3, 4, 5, 9, 1, 2] and for the second sample  [7, 9, 1, 4, 5, 6, 2] and so on...
Make sure each sample will have atleast 3 feautres/columns/attributes

* <font color='red'><b> Note - While selecting the random 60% datapoints from the whole data, make sure that the selected datapoints are all exclusive, repetition is not allowed. </b></font>

<font color='red'><b>Step - 2 </b></font>

<font color='blue'><b>Building High Variance Models on each of the sample and finding train MSE value</b></font>

*  Build a regression trees on each of 30 samples.
*  Computed the predicted values of each data point(506 data points) in your corpus.
*  Predicted house price of $i^{th}$ data point $y^{i}_{pred} =  \frac{1}{30}\sum_{k=1}^{30}(\text{predicted value of } x^{i} \text{ with } k^{th} \text{ model})$
*  Now calculate the $MSE =  \frac{1}{506}\sum_{i=1}^{506}(y^{i} - y^{i}_{pred})^{2}$

<font color='red'> <b>Step - 3 </b></font>

*  <font color='blue'><b>Calculating the OOB score </b></font>

*  Predicted house price of $i^{th}$ data point $y^{i}_{pred} =  \frac{1}{k}\sum_{\text{k= model which was buit on samples not included } x^{i}}(\text{predicted value of } x^{i} \text{ with } k^{th} \text{ model})$.
*  Now calculate the $OOB Score =  \frac{1}{506}\sum_{i=1}^{506}(y^{i} - y^{i}_{pred})^{2}$.

# <font color='red'><b>Task 2</b></font>

*  <font color='blue'><b>Computing CI of OOB Score and Train MSE</b></font>
  *   Repeat Task 1 for 35 times, and for each iteration store the Train MSE and OOB score </li>
<li> After this we will have 35 Train MSE values and 35 OOB scores </li>
<li> using these 35 values (assume like a sample) find the confidence intravels of MSE and OOB Score </li>
<li> you need to report CI of MSE and CI of OOB Score </li>
<li> Note: Refer the Central_Limit_theorem.ipynb to check how to find the confidence intravel</li>
</ol>

# <font color='red'><b>Task 3</b></font>

*  <font color='blue'><b>Given a single query point predict the price of house.</b></font>

Consider xq= [0.18,20.0,5.00,0.0,0.421,5.60,72.2,7.95,7.0,30.0,19.1,372.13,18.60] 
Predict the house price for this point as mentioned in the step 2 of Task 1.

## <font color='red'><b>A few key points</b></font>

* Remember that the datapoints used for calculating MSE score contain some datapoints that were initially used while training the base learners (the 60% sampling). This makes these datapoints partially seen (i.e. the datapoints used for calculating the MSE score are a mixture of seen and unseen data).
Whereas, the datapoints used for calculating OOB score have only the unseen data. This makes these datapoints completely unseen and therefore appropriate for testing the model's performance on unseen data.

* Given the information above, if your logic is correct, the calculated MSE score should be less than the OOB score.

* The MSE score must lie between 0 and 10.
* The OOB score must lie between 10 and 35.

* The difference between the left nad right confidence-interval values must not be more than 10. Make sure this is true for both MSE and OOB confidence-interval values.

# <font color='blue'> <b>Task - 1</b></font>

<font color='blue'><b>Step - 1</b></font>

*  <font color='blue'> <b>Creating samples</b></font>

<font color='Orange'><b>Algorithm</b></font>

![alt text](https://i.imgur.com/OfcFrUP.jpg/)

*  <font color='blue'><b> Write code for generating samples</b></font>

In [ ]:
from random import choice
def generating_samples(input_data, target_data):

    '''In this function, we will write code for generating 30 samples '''
    # you can use random.choice to generate random indices without replacement
    # Please have a look at this link https://docs.scipy.org/doc/numpy-1.16.0/reference/generated/numpy.random.choice.html for more details
    # Please follow above pseudo code for generating samples
    selecting_rows=np.random.choice(a=506,size=303,replace=False) #Selecting 303 random indices from input_data
    
    #Selecting Column Indices
    arr=np.arange(13)
    column_len= np.random.choice(a=np.arange(3,14))
    selecting_columns= np.random.choice(a=arr,size=column_len,replace=False)
    
    #Preparation of Data of 303 rows
    sample_data= input_data[selecting_rows[:,None],selecting_columns]
    target_of_sample_data=target_data[selecting_rows]

    #Replicating Data
    replacing_rows= np.random.choice(a=len(sample_data),size=203, replace=False) #Selecting 203 random indices from sampled_data
    replicated_sample_data=sample_data[replacing_rows]
    target_of_replicated_sample_data=target_of_sample_data[replacing_rows]

    #Concatinating Data
    final_sample_data=np.vstack((sample_data,replicated_sample_data))
    final_target_data=np.vstack((target_of_sample_data.reshape(-1,1),target_of_replicated_sample_data.reshape(-1,1)))

    return final_sample_data, final_target_data, selecting_rows, selecting_columns  

<font color='cyan'> <b> Grader function - 1 </b> </fongt>

In [ ]:
def grader_samples(a,b,c,d):
    length = (len(a)==506  and len(b)==506)
    sampled = (len(a)-len(set([str(i) for i in a]))==203)
    rows_length = (len(c)==303)
    column_length= (len(d)>=3)
    assert(length and sampled and rows_length and column_length)
    return True
a,b,c,d = generating_samples(x, y)
grader_samples(a,b,c,d)

True

*  <font color='blue'> <b>Create 30 samples </b>

![alt text](https://i.imgur.com/p8eZaWL.jpg)

In [ ]:
# Use generating_samples function to create 30 samples 
# store these created samples in a list
list_input_data =[]
list_output_data =[]
list_selected_row= []
list_selected_columns=[]

for i in range(0,30):
  a,b,c,d=generating_samples(x,y)
  list_input_data.append(a)
  list_output_data.append(b)
  list_selected_row.append(c)
  list_selected_columns.append(d)

NameError: ignored

<font color='cyan'> <b>Grader function - 2 </b></font>

In [ ]:
def grader_30(a):
    assert(len(a)==30 and len(a[0])==506)
    return True
grader_30(list_input_data)

True

<font color='red'><b>Step - 2 </b></font>

<font color='orange'><b>Flowchart for building tree</b></font>

![alt text](https://i.imgur.com/pcXfSmp.png)

*  <font color='blue'><b> Write code for building regression trees</b></font>

In [ ]:
from sklearn.tree import DecisionTreeRegressor
list_of_all_models=[]

for i in range(len(list_input_data)):
  model=DecisionTreeRegressor(max_depth=None)
  regressor=model.fit(list_input_data[i],list_output_data[i])
  list_of_all_models.append(regressor)

<font color='orange'><b>Flowchart for calculating MSE </b></font>

![alt text](https://i.imgur.com/sPEE618.png)

After getting predicted_y for each data point, we can use sklearns mean_squared_error to calculate the MSE between predicted_y and actual_y.

*  <font color='blue'><b> Write code for calculating MSE</b></font>

In [ ]:
y_pred_overall=[] #To store predicted values of y of all 506 datapoints of x.
for i in range(len(x)): #For each datapoint in x
  array_of_Y=[] #To store all values of y_pred of each of the 30 models, of each datapoint
  for j in range(len(list_selected_columns)): #Range 30
    p=list_of_all_models[j].predict(x[i][list_selected_columns[j]].reshape(1,-1)) #Calculating y_pred
    array_of_Y.append(p[0])
  predicted_price= np.median(array_of_Y) #Median of predicted price of each of the 30 models
  y_pred_overall.append(predicted_price)
print("Predicted Ys= ",y_pred_overall[0:10])
print('_'*100)
print("Actual Ys: ", y[0:10])
print('_'*100)

#Calculation of Mean Square Error:
square_of_diff=[]
for i in range(len(y)):
  diff_sq=(y[i]-y_pred_overall[i])**2
  square_of_diff.append(diff_sq)
MSE= np.round(sum(square_of_diff)/506,4)
print("Mean Square Error= ",MSE)
RMSE=(MSE)**1/2
print("Root Mean Square Error= ",RMSE)

Predicted Ys=  [24.0, 21.6, 34.7, 33.4, 35.68333333333334, 28.7, 22.9, 27.1, 16.5, 18.9]
____________________________________________________________________________________________________
Actual Ys:  [24.  21.6 34.7 33.4 36.2 28.7 22.9 27.1 16.5 18.9]
____________________________________________________________________________________________________
Mean Square Error=  0.1114
Root Mean Square Error=  0.0557


<font color='blue'><b>Step - 3 </b></font>

<font color='orange'><b>Flowchart for calculating OOB score</b></font>

![alt text](https://i.imgur.com/95S5Mtm.png)

*  <font color='blue'><b> Write code for calculating OOB score </b></font>

In [ ]:
y_pred_overall_oob=[] #To Store Predicted values of each Datapoint
for i in range(len(x)): #In Range 506
  array_of_Y_oob=[] #To Store predicted Y of each Datapoint which is not present while training Model_i.
  for j in range(len(list_input_data)): #Range 30
    if x[i] not in list_input_data[j]: #If a datapoint is not present in the training data given to Model_i.
      p_oob=list_of_all_models[j].predict(x[i][list_selected_columns[j]].reshape(1,-1)) #Y_predict of datapoint x[i]
      array_of_Y_oob.append(p_oob[0])
  predicted_price_oob= np.median(array_of_Y_oob) #Predicted price= Median predicted price of all 30 models.
  y_pred_overall_oob.append(predicted_price_oob)

In [ ]:
print("Predicted Ys= ",y_pred_overall_oob[0:10])
print('_'*100)
print("Actual Ys: ", y[0:10])
print('_'*100)



Predicted Ys=  [24.0, 21.6, 34.7, 34.05, 36.2, 28.7, 22.9, 27.1, 16.5, 18.9]
____________________________________________________________________________________________________
Actual Ys:  [24.  21.6 34.7 33.4 36.2 28.7 22.9 27.1 16.5 18.9]
____________________________________________________________________________________________________


Now calculate the $OOB Score =  \frac{1}{506}\sum_{i=1}^{506}(y^{i} - y^{i}_{pred})^{2}$.

In [ ]:
#Calculation of OOB Score:
square_of_diff_oob=[]
for i in range(len(y)):
  diff_sq_oob=(y[i]-y_pred_overall_oob[i])**2
  square_of_diff_oob.append(diff_sq_oob)
OOB_Score= np.round(sum(square_of_diff_oob)/506,4)
print("OOB Score= ",OOB_Score)
print('_'*100)

OOB Score=  0.0906
____________________________________________________________________________________________________


# <font color='blue'><b>Task 2</b></font>

In [ ]:
list_of_MSE,list_of_OOBS=[],[]

for num in range(35):
  list_input_data =[]
  list_output_data =[]
  list_selected_row= []
  list_selected_columns=[]
  
  for n in range(0,30):
    a,b,c,d=generating_samples(x,y)
    list_input_data.append(a)
    list_output_data.append(b)
    list_selected_row.append(c)
    list_selected_columns.append(d)

  list_of_all_models=[]
  for i in range(len(list_input_data)):
    model=DecisionTreeRegressor(max_depth=None)
    regressor=model.fit(list_input_data[i],list_output_data[i])
    list_of_all_models.append(regressor)

  y_pred_overall=[] 
  for i in range(len(x)): 
    array_of_Y=[]
    for j in range(len(list_selected_columns)):
      p=list_of_all_models[j].predict(x[i][list_selected_columns[j]].reshape(1,-1))
      array_of_Y.append(p[0])
    predicted_price= np.median(array_of_Y)
    y_pred_overall.append(predicted_price)

  #Calculation of Mean Square Error
  square_of_diff=[]
  for i in range(len(y)):
    diff_sq=(y[i]-y_pred_overall[i])**2
    square_of_diff.append(diff_sq)
  MSE= np.round(sum(square_of_diff)/506,4)
  list_of_MSE.append(MSE)

  y_pred_overall_oob=[]
  for i in range(len(x)):
    array_of_Y_oob=[]
    for j in range(len(list_input_data)):
      if x[i] not in list_input_data[j]:
        p_oob=list_of_all_models[j].predict(x[i][list_selected_columns[j]].reshape(1,-1))
        array_of_Y_oob.append(p_oob[0])
    predicted_price_oob= np.median(array_of_Y_oob)
    y_pred_overall_oob.append(predicted_price_oob)

  #Calculation of OOB Score:
  square_of_diff_oob=[]
  for i in range(len(y)):
    diff_sq_oob=(y[i]-y_pred_overall_oob[i])**2
    square_of_diff_oob.append(diff_sq_oob)
  OOB_Score= np.round(sum(square_of_diff_oob)/506,4)
  list_of_OOBS.append(OOB_Score)

In [ ]:
print("MSE List= ",list_of_MSE)
print('-'*50)
print("Length of MSE List= ", len(list_of_MSE))
print("_"*100)
print("OOBS List= ",list_of_OOBS)
print('-'*50)
print("Length of OOBS List= ", len(list_of_OOBS))


MSE List=  [0.2609, 0.0827, 0.0751, 0.0501, 0.121, 0.0152, 0.0478, 0.0522, 0.0989, 0.0686, 0.0258, 0.0178, 0.3679, 0.0587, 0.0565, 0.0275, 0.0342, 0.0035, 0.1791, 0.0419, 0.0681, 0.0501, 0.0284, 0.0674, 0.1005, 0.1428, 0.044, 0.062, 0.0451, 0.0893, 0.0127, 0.0948, 0.1784, 0.0791, 0.2601]
--------------------------------------------------
Length of MSE List=  35
____________________________________________________________________________________________________
OOBS List=  [0.2667, 0.0844, 0.0743, 0.0712, 0.1149, 0.0218, 0.0423, 0.0527, 0.091, 0.1194, 0.0156, 0.0636, 0.7587, 0.1092, 0.068, 0.0331, 0.0615, 0.0214, 0.4943, 0.0827, 0.1427, 0.0729, 0.0769, 0.1714, 0.0486, 0.0956, 0.0247, 0.0921, 0.1376, 0.0445, 0.0099, 0.0933, 0.1593, 0.1055, 0.3226]
--------------------------------------------------
Length of OOBS List=  35


### Population Means of MSE and OOBS


In [ ]:
from statistics import mean
from statistics import stdev
population_mean_mse= np.round(mean(list_of_MSE),3)
population_mean_oobs= np.round(mean(list_of_OOBS),3)
print("Population Mean of MSEs= ",population_mean_mse)
print("Population Mean of OOBS= ",population_mean_oobs)

population_std_dev_MSE = np.round(stdev(list_of_MSE),3)
population_std_dev_OOBS= np.round(stdev(list_of_OOBS),3)
print("Population Standard Deviation of MSEs= ",population_std_dev_MSE)
print("Population Standard Deviation of OOBS= ",population_std_dev_OOBS)

Population Mean of MSEs=  0.086
Population Mean of OOBS=  0.121
Population Standard Deviation of MSEs=  0.079
Population Standard Deviation of OOBS=  0.145


### Finding Confidence Intervals
a) **MSE**




* From **Central Limit Theorem** we know that the Sampling Distribution of Sample means is approximately equal to a Normal/Gaussian Distribution with **Mean=Population Mean** and **Std-Error= Population Standard Deviation/(Number of Samples)1/2**. The larger the number of samples and sample size, the more closer it is to a Normal Distribution.

* Here we have only 35 MSEs. Hence we can take the whole population as a sample and calculate sample mean (No need of further sampling).
And also from **66,95,99.7 rule** we know that between the range **[mu_x-(2*Std_Err), mu_x+(2*Std_Err)]**, there lies **95%** of the Data. Hence we can calculate the Confidence Interval of the Mean using the above.

<font size='3'>From the above equations Let us construct an interval = [<font color='red'> $\overline{x}$- 2\*$\frac{\sigma}{\sqrt{n}}$, $\overline{x}$+2\*$\frac{\sigma}{\sqrt{n}}$</font>]</font>, 
where <font color='red'> $\overline{x}$</font> is the sample mean and  <font color='red'> $\frac{\sigma}{\sqrt{n}}$ </font> is the standard error.

In [ ]:
sample_size_mse=len(list_of_MSE) #Taking the whole list as a sample
sample_mean_MSE = np.round(mean(list_of_MSE),3) 
sample_std_dev_MSE= np.round(stdev(list_of_MSE),3)
std_err_MSE= np.round(sample_std_dev_MSE/np.sqrt(sample_size_mse),3) #Standard Error
Left_lim_MSE= np.round((sample_mean_MSE- 2*(std_err_MSE)),3) #Left Limit
Right_lim_MSE= np.round((sample_mean_MSE+ 2*(std_err_MSE)),3) #Right Limit
print("Sample Mean of MSEs: ",sample_mean_MSE )
print("The 95% Confidence Interval of MSEs: ", [Left_lim_MSE,Right_lim_MSE])

Sample Mean of MSEs:  0.086
The 95% Confidence Interval of MSEs:  [0.06, 0.112]


Confidence Interval with 95% Confidence implies that the Population Mean of 'Mean Square Errors' lies in the above interval **95% of the time.**

Hence 
Confidence Interval of **Mean Square Errors** with **95%** Confidence=  **[0.06, 0.112]**

b) **OOBS**


In [ ]:
sample_size_oobs=len(list_of_OOBS) 
sample_mean_OOBS = np.round(mean(list_of_OOBS),3) 
sample_std_dev_OOBS= np.round(stdev(list_of_OOBS),3)
std_err_OOBS= np.round(sample_std_dev_OOBS/np.sqrt(sample_size_oobs),3) #Standard Error
Left_lim_OOBS= np.round((sample_mean_OOBS- 2*(std_err_OOBS)),3) #Left Limit
Right_lim_OOBS= np.round((sample_mean_OOBS+ 2*(std_err_OOBS)),3) #Right Limit
print("Sample Mean of OOBS: ",sample_mean_OOBS )
print("The 95% Confidence Interval of OOBS: ", [Left_lim_OOBS,Right_lim_OOBS])

Sample Mean of OOBS:  0.121
The 95% Confidence Interval of OOBS:  [0.071, 0.171]


Hence 
Confidence Interval of **OOBS Scores** with **95%** Confidence=  **[0.071, 0.171]**

In [ ]:
from prettytable import PrettyTable
x = PrettyTable()
x = PrettyTable(["For","No: of Samples", "Sample Size", "Sample Mean","Sample Standard Dev", "Std_Error"+u"\u2248"+"Sample_std_Dev/"+u"\u221A"+"n","C.I"])
#MSE
row=["MSE List",1,35]
row.append(sample_mean_MSE)
row.append(sample_std_dev_MSE)
row.append(std_err_MSE)
row.append([Left_lim_MSE,Right_lim_MSE])
x.add_row(row)

#OOBS
row1=["OOBS List",1,35]
row1.append(sample_mean_OOBS)
row1.append(sample_std_dev_OOBS)
row1.append(std_err_OOBS)
row1.append([Left_lim_OOBS,Right_lim_OOBS])
x.add_row(row1)
print(x)

+-----------+----------------+-------------+-------------+---------------------+-----------------------------+----------------+
|    For    | No: of Samples | Sample Size | Sample Mean | Sample Standard Dev | Std_Error≈Sample_std_Dev/√n |      C.I       |
+-----------+----------------+-------------+-------------+---------------------+-----------------------------+----------------+
|  MSE List |       1        |      35     |    0.086    |        0.079        |            0.013            | [0.06, 0.112]  |
| OOBS List |       1        |      35     |    0.121    |        0.145        |            0.025            | [0.071, 0.171] |
+-----------+----------------+-------------+-------------+---------------------+-----------------------------+----------------+


# <font color='blue'><b>Task 3</b></font>

<font color='orange'><b>Flowchart for Task 3</b></font>

<b>Hint: </b> We created 30 models by using 30 samples in TASK-1. Here, we need send query point "xq"  to 30 models and perform the regression on the output generated by 30 models.

![alt text](https://i.imgur.com/Y5cNhQk.png)

*  <font color='blue'><b> Write code for TASK 3 </b></font>

In [ ]:
xq=np.array([0.18,20.0,5.00,0.0,0.421,5.60,72.2,7.95,7.0,30.0,19.1,372.13,18.60])
arr_of_Y=[]
for i in range(len(list_selected_columns)):
  y_predict=list_of_all_models[i].predict(xq[list_selected_columns[i]].reshape(1,-1))
  arr_of_Y.append(y_predict[0])
#Predicted y of Datapoint xq=
y=np.median(arr_of_Y)
print("Predicted House Price for the above Datapoint xq\n y= ",y)

Predicted House Price for the above Datapoint xq
 y=  18.7


<font color='red'><b>Write observations for task 1, task 2, task 3 indetail</b></font>

# OBSERVATIONS:

The Boston House Price Dataset is one which contains the Prices of Houses in Boston as the Response Variable and 13 Features(Numerical) as the Independent Variable.
It contains 506 Datapoints (**Shape=506*13**). Here we have been given the preprocessed Data.

## Task 1
In Task-1 for **Boston House Price Dataset** , we have applied **DecisionTreeRegression** using **Random Forest**. We have randomly sampled rows and columns from the original dataset and created **30 sample datasets**. We have applied DecisionTreeRegression() on each of the 30 samples. For each of the 506 datapoints in the original dataset, we predicted the House Price using each of the 30 models and found out the mean House Price as Y_pred.We then calculated Mean Square Error and Out of Box Score(OOBS) Score for the whole Dataset.
* 1. **Mean Square Error(MSE)** (between Actual Ys and Predicted Ys)= **0.0765**
  2. **Root Mean Square Error**= **0.03825**
* **OOBS(Out of Box Score)**= **0.0906**

## Task 2
In Task 2, we trained all the Datapoints for each of the 30 models for **35 times**. We created a list of MSEs and OOBS, containing 35 values each.
We found the 95% Confidence Interval of the MSEs and OOBs.

* **95%** C.I of **MSEs**=[**0.06, 0.112**]
* **95%** C.I of **OOBS**=[**0.071, 0.171**]

## Task 3

In Task 3, we found out the House Price for a given Datapoint xq=[0.18,20.0,5.00,0.0,0.421,5.60,72.2,7.95,7.0,30.0,19.1,372.13,18.60]. We trained this Datapoint on each of the 30 Models and found out the Median House Price as
**y=18.7**.

